indigo_app/views/tasks.py

```diff
@@ -398,6 +411,38 @@ def get_redirect_url(self):
         return reverse('task_detail', kwargs={'place': self.kwargs['place'], 'pk': self.kwargs['pk']})
 
 
+class TaskChangeBlockingTasksView(SingleTaskViewBase, View, SingleObjectMixin):
+    # permissions
+    permission_required = ('indigo_api.change_task',)
+
+    http_method_names = ['post']
+
+    def post(self, request, *args, **kwargs):
+        task = self.get_object()
+        user = self.request.user
+        task.updated_by_user = user
+        ids = self.request.POST.getlist('blocked_by')
+
+        if ids:
+            blocked_by = Task.objects.filter(id__in=ids).all()
+        else:
+            blocked_by = []
+
+        task.blocked_by.set(blocked_by)
+        if blocked_by and has_transition_perm(task.block, self):
```

Comment:
We're only checking the perm *after* the task has been blocked. Instead, check the perm first and then pass the user and tasks to block on into task.block(), like we do with task.close(), then that method does the business logic.

Reply:
If I'm understanding you right, you're suggesting that the block transition method on task set the blocked_by tasks on our task in the process of blocking it (and unblock would clear them). 

We also need to make provision for the blocked_by list just being updated without there being a change in state; that can only happen in the view unless I'm mistaken.

So `task.blocked_by.set(blocked_by)` would only be called in the case of there being blocked_by tasks but the transition permission being False (because you can't move from blocked into blocked).

Reply:
Hmm, good point - maybe get around that by allowing blocked -> blocked as a state transition? Otherwise, set the blocked tasks outside of task.blocked(). Otherwise it's duplicate code.

Reply:
I don't love a blocked-to-blocked transition, and we'd have to then special-case the actions on that so that we don't end up with multiple 'user blocked task' actions in a row if they're just updating the list, so I've changed it to set the blocking tasks outside of the block and unblock methods

Reply:
Just sanity check the if tree... what if the task is blocked by the user doesn't have necessary perms to make this change?

Reply:
Ah. I thought the permission_required as well as disabling the button if the user doesn't have permissions would cover us, but I see the country permission isn't checked (or it's not obvious if it is, maybe I've missed it now).

So because indigo_api.change_task is listed under permission_required, is the only other thing I need to check for the country permission? or should I explicitly check for indigo_api.change_task at the point where I'm setting the blocked_by list?

I'm not super clear on how permission_required works, but I see there's a has_permission on PlaceViewBase which checks for country permission. So should I just say `elif blocked_by and self.has_permission()`, and either raise a permission denied or go ahead and set the blocked_by list at that point?

Once I'm clear on this I can figure out how not to repeat the `set` line three times :grimacing:

Reply:
`permission_required` will be checked right at the start of the request, but assume that you don't know how `has_transition_perm` works, so you can't assume that because we've got this far the user can make this change (since has_transition_perm checks both validity and authorization, right?).

Change the `if` tree to be like this, which I think will make it clearer:

```
if blocked_by:
  # user is trying to block this task, is that transition allowed?
  if has_transition_perm(...):
    # can proceed
    ...
  else:
    raise permission denied
else:
  # user is trying to unblock, can that transition proceed?
  if has_transition_perm(..):
    # cool
    ...
  else:
    raise permission denied
```